File: okj/szoftverfejleszto/dhcp/c#/DHCP.cs
using System;
using System.Collections.Generic;
using System.IO;

var excluded = File.ReadAllLines("excluded.csv");
var reserved = ReadFileToMap("reserved.csv");
var dhcp = ReadFileToMap("dhcp.csv");

foreach(var line in File.ReadAllLines("test.csv")) {
    var split = line.Split(';');

    RunTest(split[0], split[1], dhcp, excluded, reserved);
}

using var file = new StreamWriter("dhcp_kesz.csv");

foreach(var entry in dhcp) {
    file.WriteLine(entry.Key + ';' + entry.Value);
}


Dictionary<string, string> ReadFileToMap(string file) {
    var toReturn = new Dictionary<string, string>();

    foreach(var line in File.ReadAllLines(file)) {
        var split = line.Split(';');

        toReturn.Add(split[0], split[1]);
    }

    return toReturn;
}

void RunTest(string type, string address, Dictionary<string, string> dhcp, string[] excluded, Dictionary<string, string> reserved) {
    if(type == "release"){                // Address is IP
        dhcp.Remove(dhcp[address]);
    }else{                                // Address is MAC
        if(!dhcp.ContainsKey(address)) {
            var optionalReservedIP = reserved[address];
            var ipAddress = optionalReservedIP != null ? optionalReservedIP : CreateIP(address, dhcp, excluded, reserved);

            if(!dhcp.ContainsValue(ipAddress)) {
                dhcp.Add(address, ipAddress);
            }
        }
    }
}

string CreateIP(string macAddress, Dictionary<string, string> dhcp, string[] excluded, Dictionary<string, string> reserved) {
    for(var i = 100; i <= 199; ++i) {
        var ip = "192.168.10." + i;

        if(!dhcp.ContainsValue(ip) && Array.IndexOf(excluded, ip) == -1 && !reserved.ContainsValue(ip)) {
            return ip;
        }
    }

    throw new Exception("Unable to assign IP address to mac: " + macAddress);
}

File: okj/szoftverfejleszto/dhcp/java/DHCP.java
import java.io.*;
import java.nio.file.*;
import java.util.*;

public final class DHCP {
    
    public static void main(String[] args) throws IOException {
        var excluded = Files.readAllLines(Path.of("excluded.csv"));
        var reserved = readFileToMap(Path.of("reserved.csv"));
        var dhcp = readFileToMap(Path.of("dhcp.csv"));
        
        for(var line : Files.readAllLines(Path.of("test.csv"))) {
            var split = line.split(";");
            
            runTest(split[0], split[1], dhcp, excluded, reserved);
        }
        
        try(var file = new PrintWriter("dhcp_kesz.csv")) {
            for(var entry : dhcp.entrySet()) {
                file.println(entry.getKey() + ';' + entry.getValue());
            }
        }
    }
    
    public static Map<String, String> readFileToMap(Path file) throws IOException {
        var toReturn = new HashMap<String, String>();
        
        for(var line : Files.readAllLines(file)) {
            var split = line.split(";");
            
            toReturn.put(split[0], split[1]);
        }
        
        return toReturn;
    }
    
    public static void runTest(String type, String address, Map<String, String> dhcp, List<String> excluded, Map<String, String> reserved) {
        if(type.equals("release")){           // Address is IP
            dhcp.values().remove(address);
        }else{                                // Address is MAC
            if(!dhcp.containsKey(address)) {
                var optionalReservedIP = reserved.get(address);
                var ipAddress = optionalReservedIP != null ? optionalReservedIP : createIP(address, dhcp, excluded, reserved);
                
                if(!dhcp.containsValue(ipAddress)) {
                    dhcp.put(address, ipAddress);
                }
            }
        }
    }
    
    public static String createIP(String macAddress, Map<String, String> dhcp, List<String> excluded, Map<String, String> reserved) {
        for(var i = 100; i <= 199; ++i) {
            var ip = "192.168.10." + i;
            
            if(!dhcp.containsValue(ip) && !excluded.contains(ip) && !reserved.containsValue(ip)) {
                return ip;
            }
        }
        
        throw new IllegalStateException("Unable to assign IP address to mac: " + macAddress);
    }
}